bpf/flows.c

```diff
@@ -95,59 +151,137 @@ static inline int fill_ip6hdr(struct ipv6hdr *ip, void *data_end, struct flow *f
+    flow_id->eth_protocol = __bpf_ntohs(eth->h_proto);
+
+    if (flow_id->eth_protocol == ETH_P_IP) {
         struct iphdr *ip = (void *)eth + sizeof(*eth);
-        return fill_iphdr(ip, data_end, flow);
-    } else if (flow->protocol == ETH_P_IPV6) {
+        return fill_iphdr(ip, data_end, flow_id, flags);
+    } else if (flow_id->eth_protocol == ETH_P_IPV6) {
         struct ipv6hdr *ip6 = (void *)eth + sizeof(*eth);
-        return fill_ip6hdr(ip6, data_end, flow);
+        return fill_ip6hdr(ip6, data_end, flow_id, flags);
+    } else {
+        // TODO : Need to implement other specific ethertypes if needed
+        // For now other parts of flow id remain zero
+        memset (&(flow_id->src_ip),0, sizeof(struct in6_addr));
+        memset (&(flow_id->dst_ip),0, sizeof(struct in6_addr));
+        flow_id->protocol = 0;
+        flow_id->src_port = 0;
+        flow_id->dst_port = 0;
     }
     return SUBMIT;
 }
 
-// parses flow information for a given direction (ingress/egress)
-static inline int flow_parse(struct __sk_buff *skb, u8 direction) {
+static inline void export_flow_id (flow *key, flow_id_v id, u8 direction) {
+    key->protocol = id.eth_protocol;
+    key->direction = direction;
+    __builtin_memcpy(key->data_link.src_mac, id.src_mac, ETH_ALEN);
+    __builtin_memcpy(key->data_link.dst_mac, id.dst_mac, ETH_ALEN);
+    key->network.src_ip = id.src_ip;
+    key->network.dst_ip = id.dst_ip;
+    key->transport.src_port = id.src_port;
+    key->transport.dst_port = id.dst_port;
+    key->transport.protocol = id.protocol;
+}
 
+static inline int flow_monitor (struct __sk_buff *skb, u8 direction) {
     // If sampling is defined, will only parse 1 out of "sampling" flows
     if (sampling != 0 && (bpf_get_prandom_u32() % sampling) != 0) {
         return TC_ACT_OK;
     }
-
-    void *data = (void *)(long)skb->data;
     void *data_end = (void *)(long)skb->data_end;
+    void *data = (void *)(long)skb->data;
+    flow_id_v id;
+    int rc = TC_ACT_OK;
+    u32 flags = 0;
+    flow_record *flow_event;
+    flow_metrics *my_flow_counters;
+    __u64 current_time = bpf_ktime_get_ns();
 
-    struct flow *flow = bpf_ringbuf_reserve(&flows, sizeof(struct flow), 0);
-    if (!flow) {
+    struct ethhdr *eth = data;
+    if (fill_ethhdr(eth, data_end, &id, &flags) == DISCARD) {
         return TC_ACT_OK;
     }
+    if (direction == INGRESS) {
+        my_flow_counters = bpf_map_lookup_elem(&xflow_metric_map_ingress, &id);
+    } else {
+        my_flow_counters = bpf_map_lookup_elem(&xflow_metric_map_egress, &id);
+    }
 
-    struct ethhdr *eth = data;
-    if (fill_ethhdr(eth, data_end, flow) == DISCARD) {
-        bpf_ringbuf_discard(flow, 0);
+    if (my_flow_counters != NULL) {
+        my_flow_counters->packets += 1;
+        my_flow_counters->bytes += skb->len;
+        my_flow_counters->last_pkt_ts = current_time;
+        if (flags & TCP_FIN_FLAG || flags & TCP_RST_FLAG) {
```

Comment:
Is it ok for long lasting connections? or should we add other eviction criteria, like what exists in ovs/ipfix, based on a packets limit reached, and/or a timeout ?

Reply:
Currently, I have implemented a timeout logic for flows to be evicted from userspace. Please Refer [tracer.go](https://github.com/netobserv/netobserv-ebpf-agent/pull/39/files#diff-f3248f855cc319b6d96125842a513586d6d8cbfb1e241a93e79911d08e41c728). We can potentially add other eviction logic there.

Reply:
It seems to me that yes, it's necessary to have also the timeout/max-size eviction mechanism, because else data could be retained on the cpu maps potentially indefinitely, regardless the cache timeout/max-size configured. (Unless there is a mechanism that allows flushing kernel-space maps triggered from the user-space, but I didn't see that, right?)

Or maybe there's something I don't understand, I'm not yet super familiar with the agent code

Reply:
I am currently using FIN/RST flag to communicate to the userspace agent that, the flow is complete. This will trigger the userspace agent to evict the entries of the flows immediately, instead of waiting for the timeout. 
This is done mainly to evict entries immediately instead of waiting until evictionTimeout to clear the map space for other flows.

Reply:
As long as our product is not (yet) connection aware, I think this is not needed as it won't provide any extra information/input to the users. We actually follow the "standard" definition of flow that do not consider the connection status: https://datatracker.ietf.org/doc/html/rfc3954#section-2

> a set of IP packets passing an Observation Point in the network during a certain time interval.

The code it's correct but feel free to remove it if you think it can clean/simplify the code (I'd do it)